### PR TITLE
Async Recursive URL loader

### PR DESCRIPTION
**Pull request by finger-bone**
Description: This PR improves the function of recursive_url_loader, such as limiting the depth of the access, and customizable extractors(from the raw webpage to the text of the Document object), so that users can use other tools to extract the webpage. This PR also includes the document and test for the new loader.
Old PR closed due to project structure change. #7756

Because socket requests are not allowed, the old unit test was removed.
Issue: N/A
Dependencies: asyncio, aiohttp
Tag maintainer: @rlancemartin
Twitter handle: @ Zend_Nihility

### CODE REVIEW

**Comment by rlancemartin** [comment]
@proximal-phalanx can we add back the test to this PR?

**Comment by rlancemartin** [comment]
@hinthornw here is updated PR.

@eyurtsev in case you want to have a look, too.

i just merged master, and ran ntbk.

i asked @proximal-phalanx to add back test from earlier PR (https://github.com/langchain-ai/langchain/pull/7756).

**Review by rlancemartin — COMMENTED** [review]
Merged master and can confirm notebook examples work as expected.

@hinthornw may test w/ additional test urls. 

@proximal-phalanx can add back the test, which we need.

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 1]
```diff
@@ -1,11 +1,30 @@
...
+import re
+from typing import Callable, Iterator, List, Optional, Set, Union
 from urllib.parse import urljoin, urlparse
 
 import requests
 
 from langchain.docstore.document import Document
 from langchain.document_loaders.base import BaseLoader
 
+try:
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:10]: We shouldn't be printing on the outer level here - would want to include when we initialize the class only

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 2]
```diff
@@ -1,11 +1,30 @@
...
+    import aiohttp
+except ImportError:
+    print("The aiohttp package is required for the RecursiveUrlLoader.")
+    print("Please install it with `pip install aiohttp`.")
+
+try:
+    import asyncio
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:17]: This is python  stdlib so no need to try/catch
https://docs.python.org/3/library/asyncio.html

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 3]
```diff
@@ -14,19 +33,114 @@ def __init__(
         self,
         url: str,
         exclude_dirs: Optional[str] = None,
+        use_async: bool = False,
+        extractor: Callable[[str], str] = lambda x: x,
+        max_depth: int = 2,
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:38]: I would expect to default to None

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 4]
```diff
@@ -14,19 +33,114 @@ def __init__(
...
+            base_url (str): the base url of the html
+
+        Returns:
+            List[str]: sub links
+        """
+        # Get all links that are relative to the root of the website
+        all_links = re.findall(r"href=[\"\'](.*?)[\"\']", raw_html)
+        absolute_paths = []
+        # Process the links
+        for link in all_links:
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:78]: The top checks could be simplified:


```python
    invalid_prefixes = ("javascript:", "mailto:", "#")
    invalid_suffixes = (".css", ".js", ".ico", ".png", ".jpg", ".jpeg", ".gif", ".svg")

    for link in all_links:
        # Ignore blacklisted patterns
        if link.startswith(invalid_prefixes) or link.endswith(invalid_suffixes):
            continue
```

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 5]
```diff
@@ -14,19 +33,114 @@ def __init__(
...
+            ]
+            if any([link.endswith(ext) for ext in invalid_ext]):
+                continue
+            # Some may be absolute links like https://to/path
+            if link.startswith("http"):
+                if (not self.prevent_outside) or (
+                    self.prevent_outside and link.startswith(base_url)
+                ):
+                    absolute_paths.append(link)
+            else:
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:105]: This always adds relative paths then?

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py [hunk 6]
```diff
@@ -14,19 +33,114 @@ def __init__(
...
+                absolute_paths.append(urljoin(base_url, link))
+                continue
+            # Some may have omitted the protocol like //to/path
+            if link.startswith("//"):
+                absolute_paths.append(f"{urlparse(base_url).scheme}:{link}")
+                continue
+        # Remove duplicates, also do a filter to prevent outside links
+        absolute_paths = list(
+            set(
+                filter(
```
Comment (hinthornw) [inline libs/langchain/langchain/document_loaders/recursive_url_loader.py:119]: Would be more pythonic to do something like

```
    absolute_paths = list(set([path for path in absolute_paths if not self.prevent_outside or path.startswith(base_url) and path != base_url]))

```

**Review by hinthornw — COMMENTED** [review]
Have you checked that this works on a number of different examples?

File: libs/langchain/langchain/document_loaders/recursive_url_loader.py (hunk 5 above)
Comment (proximal-phalanx) [reply libs/langchain/langchain/document_loaders/recursive_url_loader.py:83]: Yes, relative paths are certainly sub-domains of the target URL, so it's our target.

**Comment by proximal-phalanx** [comment]
@hinthornw I have modified the code according to your review of the new commit. 
The test is also added, but the test on GitHub prevents the usage of the socket(requests and other network IO libraries) so I'm not sure if it works here. The test is successful if run on the local machine with pytest.
I have tested it on other websites myself and it works fine.

**Comment by rlancemartin** [comment]
Thanks!

**Comment by rlancemartin** [comment]
> Have you checked that this works on a number of different examples?

@proximal-phalanx you have test case and the two examples in the notebook; any additional examples that you've looked at?

**Comment by rlancemartin** [comment]
@proximal-phalanx pls run make format to fix lint errors.

**Comment by rlancemartin** [comment]
Thanks! I can test some additional links tonight and merge if all is well.

**Comment by proximal-phalanx** [comment]
> @proximal-phalanx pls run make format to fix lint errors.

Fixed, it was a comment that was too long., lint on my machine didn't detect it, somehow.

> @proximal-phalanx you have test case and the two examples in the notebook; any additional examples that you've looked at?

I have also tried this on cpp reference, https://en.cppreference.com/w/, npm documentation, https://docs.npmjs.com/, the register news, https://www.theregister.com, world health organization, https://www.who.int, all working as expected.

**Comment by proximal-phalanx** [comment]
Btw, not really sure why the core test failed? There is not even a single prompt to tell me where the error happened, it just said plainly error. If something could go wrong in the test, it could be the usage of network io.

**Comment by rlancemartin** [comment]
> Btw, not really sure why the core test failed? There is not even a single prompt to tell me where the error happened, it just said plainly error. If something could go wrong in the test, it could be the usage of network io.

Interesting. I can't reproduce the test failure if I run it in a notebook:
```
url = "https://docs.python.org/3.9/"
loader = RecursiveUrlLoader(
    url=url, extractor=lambda _: "placeholder", use_async=True, max_depth=1
)
docs = loader.load()
len(docs)
```

Does return 24 as expected by the test.

Problem is probably `--disable-socket` option when we run tests:

```
poetry run pytest --disable-socket --allow-unix-socket tests/unit_tests/
```

This option will disable all socket connections during the test, which would prevent the RecursiveUrlLoader from making HTTP requests to load the URLs.

**Comment by rlancemartin** [comment]
The test appears to pass if we remove `--disable-socket`:
```
poetry run pytest --allow-unix-socket tests/unit_tests/
```
But there are other issues, so it need some deeper attention:
```
4.61s call     tests/unit_tests/retrievers/test_tfidf.py::test_from_texts
4.01s call     tests/unit_tests/llms/test_openai.py::test_openai_async_retries
4.01s call     tests/unit_tests/llms/test_openai.py::test_openai_retries
3.09s call     tests/unit_tests/document_loaders/test_recursive_url_loader.py::test_sync_recursive_url_loader
1.88s call     tests/unit_tests/vectorstores/test_sklearn.py::test_sklearn
============================================================ short test summary info =============================================================
FAILED tests/unit_tests/test_pytest_config.py::test_socket_disabled - Failed: DID NOT RAISE <class 'pytest_socket.SocketBlockedError'>
============================================ 1 failed, 1023 passed, 85 skipped, 40 warnings in 33.80s ============================================
(llama) rlm@Rs-MacBook-Pro langchain % poetry run pytest --allow-unix-socket tests/unit_tests/
```

IMO, we get this in w/o the test to un-break the loader.

Then, I'll discuss w/ @baskaryan about how we can modify tests to enable socket connections.

**Comment by rlancemartin** [comment]
> The test appears to pass if we remove `--disable-socket`:
> 
> ```
> poetry run pytest --allow-unix-socket tests/unit_tests/
> ```
> 
> But there are other issues, so it need some deeper attention:
> 
> ```
> 4.61s call     tests/unit_tests/retrievers/test_tfidf.py::test_from_texts
> 4.01s call     tests/unit_tests/llms/test_openai.py::test_openai_async_retries
> 4.01s call     tests/unit_tests/llms/test_openai.py::test_openai_retries
> 3.09s call     tests/unit_tests/document_loaders/test_recursive_url_loader.py::test_sync_recursive_url_loader
> 1.88s call     tests/unit_tests/vectorstores/test_sklearn.py::test_sklearn
> ============================================================ short test summary info =============================================================
> FAILED tests/unit_tests/test_pytest_config.py::test_socket_disabled - Failed: DID NOT RAISE <class 'pytest_socket.SocketBlockedError'>
> ============================================ 1 failed, 1023 passed, 85 skipped, 40 warnings in 33.80s ============================================
> (llama) rlm@Rs-MacBook-Pro langchain % poetry run pytest --allow-unix-socket tests/unit_tests/
> ```
> 
> IMO, we get this in w/o the test to un-break the loader.
> 
> Then, I'll discuss w/ @baskaryan about how we can modify tests to enable socket connections.

@proximal-phalanx can you please open a separate PR w/ the test?

I'll then discuss w/ @baskaryan. 

In the meantime, im going to get get this in (assuming tests pass, which looks like they will).

url testing includes:
+ https://js.langchain.com/docs/modules/memory/integrations/
+ https://docs.python.org/3.9/
+ cpp reference
+ https://en.cppreference.com/w/
+ npm documentation
+ https://docs.npmjs.com/
+ the register news
+ https://www.theregister.com/
+ world health organization
+ https://www.who.int/

@ai-ashok @hinthornw pls use and report any issues. of course, features like this are tricky due to large surface area of potential problems from different urls, but at least we'll have a test to catch regressions. as for what caused the recent regression, the only obvious PR that i see is from @rysloan4 https://github.com/langchain-ai/langchain/pull/8582 but it was a small change. in any case, @rysloan4 pls test this too and let us know if any issues.

**Comment by proximal-phalanx** [comment]
> @proximal-phalanx can you please open a separate PR w/ the test?
PR is here #8813 
If the network io is impossible, we could just copy-paste the HTML into the test, and use _get_child_link_sync(or async), whose parameter is the raw HTML string, to test it. Though this won't be able to test the whole loader.

**Comment by rlancemartin** [comment]
ugh, gcc error is non-obvious.

need to look at this tmrw w/ fresh eyes, so can't merge now.

@baskaryan or @hinthornw may have ideas. error reproduces (seen in twice after re-starting all tests).

https://github.com/langchain-ai/langchain/actions/runs/5774935624/job/15652621439?pr=8502

**Comment by rlancemartin** [comment]
> gcc error is non-obvious.
> 
> need to look at this tmrw w/ fresh eyes, so can't merge now.
> 
> @baskaryan or @hinthornw may have ideas. error reproduces (seen three times after re-starting all tests and pushing a fmt change).
> 
> https://github.com/langchain-ai/langchain/actions/runs/5774935624/job/15652621439?pr=8502

Seeing if this is resolved by merging master.

**Comment by rlancemartin** [comment]
@proximal-phalanx can you please merge master? Should pass now. I'm away from computer now but can approve / merge from mobile once we merge master.

**Comment by rlancemartin** [comment]
> @proximal-phalanx can you please merge master? Should pass now. I'm away from computer now but can approve / merge from mobile once we merge master.

Done and tests now pass! Im merging now cc @ai-ashok @hinthornw @rysloan4

**Pull request merged**